common/utils.c

```diff
@@ -965,3 +984,257 @@ TDNFDirName(
+        if (!dwError)
+        {
+            pr_err("Digest Update Failed\n");
+            dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+        memset(buf, 0, BUFSIZ);
+    }
+
+    if (length == -1)
+    {
+        pr_err("Metalink: validating (%s) FAILED\n", filename);
+        dwError = errno;
+        BAIL_ON_TDNF_SYSTEM_ERROR(dwError);
+    }
+
+    dwError = EVP_DigestFinal_ex(ctx, digest, &digest_length);
+    if (!dwError)
+    {
+        pr_err("Digest Final Failed\n");
+        dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+    dwError = 0;
+
+cleanup:
+    if (fd >= 0)
+    {
+        close(fd);
+    }
+    if (ctx)
+    {
+        EVP_MD_CTX_destroy(ctx);
+    }
+    return dwError;
+error:
+    goto cleanup;
+}
+
+uint32_t
+TDNFCheckHash(
+    const char *filename,
+    unsigned char *digest,
+    int type
+    )
+{
+
+    uint32_t dwError = 0;
+    uint8_t digest_from_file[EVP_MAX_MD_SIZE] = {0};
+    hash_op *hash = NULL;
+
+    if (IsNullOrEmptyString(filename) ||
+       !digest)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (type  < TDNF_HASH_MD5 || type >= TDNF_HASH_SENTINEL)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    hash = hash_ops + type;
+
+    dwError = TDNFGetDigestForFile(filename, hash, digest_from_file);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    if (memcmp(digest_from_file, digest, hash->length))
+    {
+        dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+cleanup:
+    return dwError;
+error:
+    if (!IsNullOrEmptyString(filename))
+    {
+        pr_err("Error: Validating Checksum (%s) FAILED (digest mismatch)\n", filename);
+    }
+    goto cleanup;
+}
+
+/* Returns nonzero if hex_digest is properly formatted; that is each
+   letter is in [0-9A-Za-z] and the length of the string equals to the
+   result length of digest * 2. */
+uint32_t
+TDNFCheckHexDigest(
+    const char *hex_digest,
+    int digest_length
+    )
+{
+    int i = 0;
+    if(IsNullOrEmptyString(hex_digest) ||
+       (digest_length <= 0))
+    {
+        return 0;
+    }
+    for(i = 0; hex_digest[i]; ++i)
+    {
+        if(!isxdigit(hex_digest[i]))
+        {
+            return 0;
+        }
+    }
+    return digest_length * 2 == i;
+}
+
+uint32_t
+TDNFHexToUint(
+    const char *hex_digest,
+    unsigned char *uintValue
+    )
+{
+    uint32_t dwError = 0;
+    char buf[3] = {0};
+    unsigned long val = 0;
+
+    if(IsNullOrEmptyString(hex_digest) ||
+       !uintValue)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    buf[0] = hex_digest[0];
+    buf[1] = hex_digest[1];
+
+    errno = 0;
+    val = strtoul(buf, NULL, 16);
+    if(errno)
+    {
+        pr_err("Error: strtoul call failed\n");
```

Comment:
This is correct way to log error msg. Wherever we used pr_err i was expecting like that Error: <error msg> that will help while debuging failures.